### PR TITLE
Add support for X448 and Ed448 (take 2)

### PR DESCRIPTION
**Pull request by mattcaswell**
This adds EVP layer support for X448 and Ed448 to libcrypto. TLS support is added in #5470.

This is a revised version of #5419. That PR was controversial because of the introduction of a couple of functions and how they were implemented. Those functions aren't actually necessary for the most bare-bones implementation needed for TLS support so I have excluded them from this PR. I will add them back in (possibly implemented slightly differently on the basis of feedback) in a different PR.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/ec/ecx_meth.c [hunk 1]
```diff
@@ -16,30 +16,40 @@
...
-    unsigned char pubkey[X25519_KEYLEN];
-    unsigned char *privkey;
-} X25519_KEY;
+#define X448_BITS            448
+#define ED448_BITS           456
+#define X448_SECURITY_BITS   224
+
+#define ED448_SIGSIZE        114
+
+#define ISX448(id)      ((id) == EVP_PKEY_X448 ? 1 : 0)
```
Comment (richsalz) [inline crypto/ec/ecx_meth.c]: Are the ?: really needed?

File: crypto/ec/ecx_meth.c [hunk 2]
```diff
@@ -16,30 +16,40 @@
...
+                                     : ((id) == EVP_PKEY_X448 ? X448_KEYLEN \
+                                                              : ED448_KEYLEN))
+#define KEYLEN(p)       KEYLENID((p)->ameth->pkey_id)
+
 
 typedef enum {
-    X25519_PUBLIC,
-    X25519_PRIVATE,
-    X25519_KEYGEN
+    KEY_OP_PUBLIC,
```
Comment (richsalz) [inline crypto/ec/ecx_meth.c]: Side note:  I prefer adding a name to the enum, too.

File: crypto/ec/ecx_meth.c [hunk 3]
```diff
@@ -51,69 +61,85 @@ static int ecx_key_op(EVP_PKEY *pkey, int id, const X509_ALGOR *palg,
...
-            return 0;
+            goto err;
         }
-        if (op == X25519_KEYGEN) {
-            if (RAND_bytes(xkey->privkey, X25519_KEYLEN) <= 0) {
-                OPENSSL_secure_free(xkey->privkey);
-                OPENSSL_free(xkey);
-                return 0;
+        if (op == KEY_OP_KEYGEN) {
+            if (RAND_bytes(privkey, KEYLENID(id)) <= 0) {
```
Comment (richsalz) [inline crypto/ec/ecx_meth.c]: Should use RAND_priv_bytes

File: crypto/ec/ecx_meth.c [hunk 4]
```diff
@@ -372,14 +446,36 @@ static int ecd_item_sign(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
...
                       X509_SIG_INFO_TLS);
     return 1;
 }
 
+
+static int ecd_item_sign448(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
+                         X509_ALGOR *alg1, X509_ALGOR *alg2,
+                         ASN1_BIT_STRING *str)
+{
+    /* Set algorithms identifiers */
```
Comment (richsalz) [inline crypto/ec/ecx_meth.c]: Maybe "Set algorithm identifier"

File: doc/man1/genpkey.pod [hunk 5]
```diff
@@ -241,9 +241,10 @@ numeric OID. Following parameter sets are supported:
 
 =back
 
-=head1 X25519 KEY GENERATION OPTIONS
+=head1 X25519 and X448 KEY GENERATION OPTIONS
 
-The X25519 algorithm does not currently support any key generation options.
+The X25519 and X448 algorithms do not currently support any key generation
+options.
```
Comment (richsalz) [inline doc/man1/genpkey.pod]: extra blank line before the NOTES header.

File: doc/man7/Ed25519.pod [hunk 6]
```diff
@@ -2,13 +2,15 @@
 
 =head1 NAME
 
-Ed25519 - EVP_PKEY Ed25519 support
+Ed25519,
+Ed448
```
Comment (richsalz) [inline doc/man7/Ed25519.pod]: missing dash before the text on line 7

File: doc/man7/Ed25519.pod [hunk 7]
```diff
@@ -2,13 +2,15 @@
...
+ EVP_PKEY Ed25519 and Ed448 support
 
 =head1 DESCRIPTION
 
-The B<Ed25519> EVP_PKEY implementation supports key generation, one shot
-digest sign and digest verify using PureEdDSA and B<Ed25519> (see RFC8032).
-It has associated private and public key formats compatible with
+The B<Ed25519> and B<Ed448> EVP_PKEY implementation supports key generation, one
+shot digest sign and digest verify using PureEdDSA and B<Ed25519> or B<Ed448>
```
Comment (richsalz) [inline doc/man7/Ed25519.pod]: Hyphenate one-shot

File: crypto/ec/ecx_meth.c [hunk 8]
```diff
@@ -360,7 +434,7 @@ static int ecd_item_verify(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
     return 2;
 }
 
-static int ecd_item_sign(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
+static int ecd_item_sign25519(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
                          X509_ALGOR *alg1, X509_ALGOR *alg2,
```
Comment (FdaSilvaYY) [inline crypto/ec/ecx_meth.c]: style: indent with/on previous line

File: crypto/ec/ecx_meth.c [hunk 9]
```diff
@@ -372,14 +446,36 @@ static int ecd_item_sign(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
...
-static int ecd_sig_info_set(X509_SIG_INFO *siginf, const X509_ALGOR *alg,
+static int ecd_sig_info_set25519(X509_SIG_INFO *siginf, const X509_ALGOR *alg,
                             const ASN1_STRING *sig)
 {
     X509_SIG_INFO_set(siginf, NID_undef, NID_ED25519, X25519_SECURITY_BITS,
                       X509_SIG_INFO_TLS);
     return 1;
 }
 
+
```
Comment (FdaSilvaYY) [inline crypto/ec/ecx_meth.c]: style : useful blank line ?

File: crypto/ec/ecx_meth.c [hunk 10]
```diff
@@ -372,14 +446,36 @@ static int ecd_item_sign(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
     return 3;
 }
 
-static int ecd_sig_info_set(X509_SIG_INFO *siginf, const X509_ALGOR *alg,
+static int ecd_sig_info_set25519(X509_SIG_INFO *siginf, const X509_ALGOR *alg,
                             const ASN1_STRING *sig)
```
Comment (FdaSilvaYY) [inline crypto/ec/ecx_meth.c]: Style: indent to fix

File: crypto/ec/ecx_meth.c [hunk 11]
```diff
@@ -372,14 +446,36 @@ static int ecd_item_sign(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
...
 {
     X509_SIG_INFO_set(siginf, NID_undef, NID_ED25519, X25519_SECURITY_BITS,
                       X509_SIG_INFO_TLS);
     return 1;
 }
 
+
+static int ecd_item_sign448(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
+                         X509_ALGOR *alg1, X509_ALGOR *alg2,
```
Comment (FdaSilvaYY) [inline crypto/ec/ecx_meth.c]: style: wrong indentation

File: crypto/ec/ecx_meth.c [hunk 12]
```diff
@@ -372,14 +446,36 @@ static int ecd_item_sign(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
...
+    /* Set algorithms identifiers */
+    X509_ALGOR_set0(alg1, OBJ_nid2obj(NID_ED448), V_ASN1_UNDEF, NULL);
+    if (alg2 != NULL)
+        X509_ALGOR_set0(alg2, OBJ_nid2obj(NID_ED448), V_ASN1_UNDEF, NULL);
+    /* Algorithm idetifiers set: carry on as normal */
+    return 3;
+}
+
+static int ecd_sig_info_set448(X509_SIG_INFO *siginf, const X509_ALGOR *alg,
+                            const ASN1_STRING *sig)
```
Comment (FdaSilvaYY) [inline crypto/ec/ecx_meth.c]: style: idem

File: crypto/err/openssl.txt [hunk 13]
```diff
@@ -594,13 +594,16 @@ EC_F_OSSL_ECDSA_SIGN_SIG:249:ossl_ecdsa_sign_sig
 EC_F_OSSL_ECDSA_VERIFY_SIG:250:ossl_ecdsa_verify_sig
 EC_F_PKEY_ECD_CTRL:271:pkey_ecd_ctrl
 EC_F_PKEY_ECD_DIGESTSIGN:272:pkey_ecd_digestsign
```
Comment (FdaSilvaYY) [inline crypto/err/openssl.txt]: Should this identifier (and a few others) remain after you renamed the related methods ?

File: crypto/ec/ecx_meth.c [hunk 14]
```diff
@@ -372,14 +446,36 @@ static int ecd_item_sign(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
...
+
+static int ecd_item_sign448(EVP_MD_CTX *ctx, const ASN1_ITEM *it, void *asn,
+                         X509_ALGOR *alg1, X509_ALGOR *alg2,
+                         ASN1_BIT_STRING *str)
+{
+    /* Set algorithms identifiers */
+    X509_ALGOR_set0(alg1, OBJ_nid2obj(NID_ED448), V_ASN1_UNDEF, NULL);
+    if (alg2 != NULL)
+        X509_ALGOR_set0(alg2, OBJ_nid2obj(NID_ED448), V_ASN1_UNDEF, NULL);
+    /* Algorithm idetifiers set: carry on as normal */
```
Comment (FdaSilvaYY) [inline crypto/ec/ecx_meth.c]: typo: idetifiers

File: CHANGES [hunk 15]
```diff
@@ -9,6 +9,13 @@
 
  Changes between 1.1.0g and 1.1.1 [xx XXX xxxx]
 
+  *) Added support for X448 and Ed448. Currently this is only supported in
+     libcrypto (not libssl). Use the EVP_PKEY_new_priv_key() and
+     EVP_PKEY_new_pub_key() functions to create X448 or Ed448 keys and then
```
Comment (kroeckx) [inline CHANGES]: That is out of date

File: crypto/ec/ecx_meth.c [hunk 16]
```diff
@@ -130,17 +156,18 @@ static int ecx_pub_decode(EVP_PKEY *pkey, X509_PUBKEY *pubkey)
...
-    const X25519_KEY *akey = a->pkey.ptr;
-    const X25519_KEY *bkey = b->pkey.ptr;
+    const ECX_KEY *akey = a->pkey.ecx;
+    const ECX_KEY *bkey = b->pkey.ecx;
 
     if (akey == NULL || bkey == NULL)
         return -2;
-    return !CRYPTO_memcmp(akey->pubkey, bkey->pubkey, X25519_KEYLEN);
+
+    return CRYPTO_memcmp(akey->pubkey, bkey->pubkey, KEYLEN(a)) == 0;
```
Comment (kroeckx) [inline crypto/ec/ecx_meth.c]: Should we check that they're actually the same type?

**Comment by kroeckx** [comment]
Looks good for me, other than the minor comments.

File: crypto/ec/ecx_meth.c (hunk 3 above)
Comment (kroeckx) [reply crypto/ec/ecx_meth.c]: That's actually one of my changes in #4665...

File: crypto/ec/ecx_meth.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/ec/ecx_meth.c]: I guess. I removed them

File: crypto/ec/ecx_meth.c (hunk 3 above)
Comment (mattcaswell) [reply crypto/ec/ecx_meth.c]: Fixed

File: crypto/ec/ecx_meth.c (hunk 4 above)
Comment (mattcaswell) [reply crypto/ec/ecx_meth.c]: Done

File: doc/man1/genpkey.pod (hunk 5 above)
Comment (mattcaswell) [reply doc/man1/genpkey.pod]: Fixed

File: doc/man7/Ed25519.pod (hunk 6 above)
Comment (mattcaswell) [reply doc/man7/Ed25519.pod]: Fixed

File: doc/man7/Ed25519.pod (hunk 7 above)
Comment (mattcaswell) [reply doc/man7/Ed25519.pod]: Done

File: crypto/ec/ecx_meth.c (hunk 8 above)
Comment (mattcaswell) [reply crypto/ec/ecx_meth.c]: Fixed

File: crypto/ec/ecx_meth.c (hunk 9 above)
Comment (mattcaswell) [reply crypto/ec/ecx_meth.c]: Removed

File: crypto/ec/ecx_meth.c (hunk 10 above)
Comment (mattcaswell) [reply crypto/ec/ecx_meth.c]: Fixed

File: crypto/ec/ecx_meth.c (hunk 11 above)
Comment (mattcaswell) [reply crypto/ec/ecx_meth.c]: Fixed

File: crypto/ec/ecx_meth.c (hunk 12 above)
Comment (mattcaswell) [reply crypto/ec/ecx_meth.c]: Fixed

File: crypto/err/openssl.txt (hunk 13 above)
Comment (mattcaswell) [reply crypto/err/openssl.txt]: Hmm...probably we need to do a big cleanup in one go of this kind of thing before we do the final release.

File: crypto/ec/ecx_meth.c (hunk 14 above)
Comment (mattcaswell) [reply crypto/ec/ecx_meth.c]: Fixed

File: CHANGES (hunk 15 above)
Comment (mattcaswell) [reply CHANGES]: Updated

File: crypto/ec/ecx_meth.c (hunk 16 above)
Comment (mattcaswell) [reply crypto/ec/ecx_meth.c]: No, EVP_PKEY_cmp guarantees that we will never get two values with different types.

**Comment by mattcaswell** [comment]
Fixup commits pushed addressing all comments.

**Comment by mattcaswell** [comment]
Awesome! Thanks - pushed.

**Pull request closed without merging**